src/workerd/api/queue.c++

```diff
@@ -1,36 +1,114 @@
 #include "queue.h"
+#include "src/workerd/jsg/jsg.h"
 
 #include <workerd/jsg/ser.h>
 #include <workerd/api/global-scope.h>
 #include <kj/encoding.h>
 
 namespace workerd::api {
 
-kj::Promise<void> WorkerQueue::send(
-    v8::Local<v8::Value> body, jsg::Optional<SendOptions> options, v8::Isolate* isolate) {
-  auto& context = IoContext::current();
+kj::StringPtr validateContentType(kj::StringPtr contentType) {
+  if (contentType == IncomingQueueMessage::ContentType::TEXT) {
+    return IncomingQueueMessage::ContentType::TEXT;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::OCTET_STREAM) {
+    return IncomingQueueMessage::ContentType::OCTET_STREAM;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::JSON) {
+    return IncomingQueueMessage::ContentType::JSON;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::V8) {
+    return IncomingQueueMessage::ContentType::V8;
+  }
 
-  JSG_REQUIRE(!body->IsUndefined(), TypeError, "Message body cannot be undefined");
+  JSG_FAIL_REQUIRE(TypeError, kj::str("Unsupported queue message content type: ", contentType));
+}
 
+kj::Array<kj::byte> serializeV8(jsg::Lock& js, v8::Local<v8::Value> body) {
   // Use a specific serialization version to avoid sending messages using a new version before all
   // runtimes at the edge know how to read it.
-  jsg::Serializer serializer(isolate, jsg::Serializer::Options {
+  jsg::Serializer serializer(js.v8Isolate, jsg::Serializer::Options {
     .version = 15,
     .omitHeader = false,
   });
   serializer.write(body);
-  kj::Array<kj::byte> serialized = serializer.release().data;
+  return serializer.release().data;
+}
 
-  auto client = context.getHttpClient(subrequestChannel, true, nullptr, "queue_send"_kj);
+kj::Array<kj::byte> serialize(jsg::Lock& js, v8::Local<v8::Value> body, kj::StringPtr contentType) {
+  if (contentType == IncomingQueueMessage::ContentType::TEXT) {
+    JSG_REQUIRE(
+      body->IsString(),
```

Comment:
This will prevent coercion into a string. Is that intended?

Reply:
Not necessarily, I generally expected this would only work well with strings, should I be more permissive here?

Reply:
Depends entirely on what you need. Only allowing explicit string values is fine at a minimum. If you want to allow users more flexibility, for instance, to pass in objects with custom toString implementations, then allowing for coercion is better. Really it depends on the DX you're going for.

Reply:
I see, thanks. I think only allowing strings for plain text should be fine for the moment. 
Making it more permissive later (if desired) seems fairly safe and non-breaking.

Reply:
heh, I think the most idiomatic model would to be support string coercion. But then you get weird effects if someone passes `undefined` and it comes out in the email as `"undefined"` rather than throwing an error. That's really why I think it's up to the experience you want users to have. I personally prefer keeping things fairly strict as I've had to help way too many people debug weird issues caused by coercion over the years.

Reply:
Well `undefined` in particular is covered because we explicitly disallow `undefined` via a separate check. But users could definitely hit other weird coercion bugs.